fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -562,61 +727,77 @@ Future<Void> fetchEncryptionKeys_impl(Reference<RESTKmsConnectorCtx> ctx,
 			tempStack.push(curUrl);
 
 			try {
-				std::string kmsEncryptionFullUrl = getEncryptionKeysFullUrl(ctx, curUrl->url);
-				TraceEvent("FetchEncryptionKeys_Start", ctx->uid).detail("KmsEncryptionFullUrl", kmsEncryptionFullUrl);
-				Reference<HTTP::Response> _resp =
+				std::string kmsEncryptionFullUrl = getFullRequestUrl(ctx, curUrl->url, urlSuffix);
+				TraceEvent(SevDebug, "KmsRequest_Start", ctx->uid)
+				    .detail("RequestID", requestID)
+				    .detail("FullUrl", kmsEncryptionFullUrl);
+				Reference<HTTP::Response> resp =
 				    wait(ctx->restClient.doPost(kmsEncryptionFullUrl, requestBodyRef.toString()));
-				resp = _resp;
 				curUrl->nRequests++;
 
 				try {
-					parseKmsResponse(ctx, resp, arena, outCipherKeyDetails);
+					T parsedResp = parseFunc(ctx, resp);
 
 					// Push urlCtx back on the ctx->urlHeap
 					while (!tempStack.empty()) {
 						ctx->kmsUrlHeap.emplace(tempStack.top());
 						tempStack.pop();
 					}
 
-					TraceEvent("FetchEncryptionKeys_Success", ctx->uid).detail("KmsUrl", curUrl->url);
-					return Void();
+					TraceEvent(SevDebug, "KmsRequest_Success", ctx->uid).detail("RequestID", requestID);
+					return parsedResp;
 				} catch (Error& e) {
-					TraceEvent(SevWarn, "FetchEncryptionKeys_RespParseFailure").error(e);
+					TraceEvent(SevWarn, "KmsRequest_RespParseFailure").error(e).detail("RequestID", requestID);
 					curUrl->nResponseParseFailures++;
-					// attempt to fetch encryption details from next KmsUrl
+					// attempt to do request from next KmsUrl
 				}
 			} catch (Error& e) {
-				TraceEvent("FetchEncryptionKeys_Failed", ctx->uid).error(e);
 				curUrl->nFailedResponses++;
 				if (pass > 1 && isKmsNotReachable(e.code())) {
+					TraceEvent(SevDebug, "KmsRequest_FailedUnreachable", ctx->uid)
+					    .error(e)
+					    .detail("RequestID", requestID);
 					throw e;
 				} else {
-					// attempt to fetch encryption details from next KmsUrl
+					TraceEvent(SevDebug, "KmsRequest_Error", ctx->uid).error(e).detail("RequestID", requestID);
+					// attempt to do request from next KmsUrl
 				}
 			}
 		}
 
 		if (pass == 1) {
-			// Re-discover KMS urls and re-attempt to fetch the encryption key details using newer KMS URLs
+			// Re-discover KMS urls and re-attempt request using newer KMS URLs
 			wait(discoverKmsUrls(ctx, true));
 		}
 	}
+	TraceEvent(SevDebug, "KmsRequest_Failed", ctx->uid).detail("RequestID", requestID);
 
-	// Failed to fetch encryption keys from the remote KMS
+	// Failed to do request from the remote KMS
 	throw encrypt_keys_fetch_failed();
```

Comment:
nit: given the function is repurposed to go beyond EncryptCipher usecase, maybe we want to add dedicated `kms_xx` error code; otherwise client if handling `encrypt_key_fetch_failed` in some code not related to encryption would be weird. 

We can put a TODO and handle it as a separate PR if needed.